schema/string.go

```diff
@@ -17,18 +17,30 @@ const (
 	stringUUIDVersion = 4
 )
 
-func checkStringConstraints(v string, minLength, maxLength int, t string) error {
+func checkStringConstraints(v string, minLength, maxLength int, pattern, t string) error {
 	if minLength != 0 && len(v) < minLength {
-		return fmt.Errorf("constraint check error: %s:%v %v < minimum:%v", t, v, minLength)
+		return fmt.Errorf("constraint check error: %s:%v %v < minimum:%v", t, v, len(v), minLength)
 	}
 	if maxLength != 0 && len(v) > maxLength {
-		return fmt.Errorf("constraint check error: %s:%v %v > maximum:%v", t, v, maxLength)
+		return fmt.Errorf("constraint check error: %s:%v %v > maximum:%v", t, v, len(v), maxLength)
+	}
+
+	if pattern != "" {
+		re, err := regexp.Compile(pattern)
```

Comment:
Due to performance reasons, the compilation of the regexp should occur when the field is loaded. For instance, imagine that one of us iterating at a big CSV. That would cause the same regexp to be unnecessarily compiled many times.

So, it should be an unexported field within the constraints struct, which holds the compiled regexp. At this stage, this should be only a Match. I am not sure where this initialization should go. The first place that comes to my mind is https://github.com/frictionlessdata/tableschema-go/blob/master/schema/field.go#L114. In the near future, we should create a constructor for Field struct and put together these initializations.

Another place whether we add MissingValues to Field (transformed the slice into a set to boost performance and readability): https://github.com/frictionlessdata/tableschema-go/blob/master/schema/schema.go#L49

Reply:
Yes, You're right. I need to consider with the performance.
Since I'm still not familiar with whole tableschema-go architecture.
So. I have no idea which function is the best way to convert the pattern to the internal member of constraint.
Do you have any recommend?

Reply:
Sure! 

I would suggest https://github.com/frictionlessdata/tableschema-go/blob/master/schema/field.go#L114

I've created #55 to revisit this place later on.

Reply:
ok, I still have one question about implementation.
If I compile the pattern string to regexp there (schema.go#L49), How to we handle the testing cases? 
In the testing cases, we invoke decodeString function and pass a constraint object to it.
Since the regexp member will be unexposed on the constraint field, I can't initialize that member like following way Constraints{pattern:"*.*", regexp:reg.compile(pattern)}.
Do you have any idea about this condition?
Or should I waiting for the #55 be finished ?

Reply:
> If I compile the pattern string to regexp there (schema.go#L49)

I suggested https://github.com/frictionlessdata/tableschema-go/blob/master/schema/field.go#L114

>  ,How to we handle the testing cases? n the testing cases, we invoke decodeString function and pass a constraint object to it. Since the regexp member will be unexposed on the constraint field, I can't initialize that member like following way Constraints{pattern:".", regexp:reg.compile(pattern)}. Do you have any idea about this condition?

You can do `Constraints{pattern: regexp:reg.compile(".")}`  because tests are in the same package as the structs and can access unexported struct fields.

So Constraints would be:

```go
type Constraints struct {
	// Required indicates whether this field is allowed to be null.
	// Schema.MissingValues define how the string representation can
	// represent null values.
	Required bool `json:"required,omitempty"`

	Maximum   string `json:"maximum,omitempty"`
	Minimum   string `json:"minimum,omitempty"`
	MinLength int    `json:"minLength,omitempty"`
	MaxLength int    `json:"maxLength,omitempty"`
	Pattern   string `json:"pattern,omitempty"`
        compiledRegexp *regexp.Regexp
}
```

> Or should I waiting for the #55 be finished ?

Please don't wait. I started that and realized that it is not so simple. We are going to need some amount of refactoring.